src/pudl/transform/nrelatb.py

```diff
@@ -0,0 +1,368 @@
+    """Info needed to convert a selection of a the raw NREL table into a normalized table."""
+
+    idx: list[str]
+    """Primary key columns of normalized subset table."""
+    columns: list[str]
+    """Columns reported in raw NREL table that are unique to the :attr:`idx`."""
+
+
+def transform_normlize(nrelatb: pd.DataFrame, normalizer: TableNormalizer):
+    """Normalize a subset of the NREL ATB data into a small tables."""
+    smol_table = nrelatb[normalizer.idx + normalizer.columns].drop_duplicates()
+    assert smol_table[smol_table.duplicated(normalizer.idx, keep=False)].empty
+    return smol_table.dropna(how="all", subset=normalizer.columns)
+
+
+class Normalizer(BaseModel):
+    """Class that defines how to normalize all of the NREL tables that get the :func:`transform_normlize` treatment."""
+
+    revisions: TableNormalizer = TableNormalizer(
+        idx=["report_year"],
+        columns=["revision_num", "update_date"],
+    )
+    units: TableNormalizer = TableNormalizer(
+        idx=["core_metric_parameter"],
+        columns=["units"],
+    )
+    technology_status: TableNormalizer = TableNormalizer(
+        idx=[
+            "report_year",
+            "technology_description",
+            "technology_description_detail_1",
+            "technology_description_detail_2",
+        ],
+        columns=["is_technology_mature", "is_default"],
+    )
+
+
+class TableUnstacker(BaseModel):
+    """Info needed to unstack a portion of the NREL ATB table."""
+
+    idx: list[str]
+    params: list[str]
+    """Values from the ``core_metric_parameter`` column to be included in this unstack."""
+
+    @field_validator("idx")
+    @classmethod
+    def idx_are_same_or_subset_of_idx_all(cls: Self, idx: list[str]):
+        """Are the :attr:`idx` columns either the same as or a subset of :attr:`IDX_ALL`?"""
+        assert set(idx).issubset(IDX_ALL)
+        return idx
+
+    @property
+    def idx_unstacked(self: Self) -> list[str]:
+        """Primary key columns after the table is unstacked.
+
+        All of the columns in :attr:`idx` except core_metric_parameter.
+        """
+        return [c for c in self.idx if c != "core_metric_parameter"]
+
+
+def transform_unstack(
+    nrelatb: pd.DataFrame, table_unstacker: TableUnstacker
+) -> pd.DataFrame:
+    """Generic unstacking function to convert ATB data from skinny to wider."""
+    param_mask = nrelatb.core_metric_parameter.isin(table_unstacker.params)
+    # assert not nrelatb[param_mask].duplicated(subset=table_unstacker.idx).any()
+
+    nrelatb_unstacked = (
+        nrelatb[param_mask]
+        .set_index(table_unstacker.idx)
+        .sort_index()[["value"]]
+        .drop_duplicates()
+        .unstack(level="core_metric_parameter")
+    )
+    nrelatb_unstacked.columns = nrelatb_unstacked.columns.droplevel()
+    # this next step regarding the fcr param is only applicable to
+    # the table w/ the fcr param:
+    if "fcr" in nrelatb_unstacked:
```

Comment:
Instead of having a special case within this general function, it might make sense to only call this function in the scenario table. Then we're not checking "is this the scenario table?" for all the other tables - and we also make the scenario table logic a little more explicit!


```python
def transform_unstack(
    nrelatb: pd.DataFrame, table_unstacker: TableUnstacker
) -> pd.DataFrame:
    """Generic unstacking function to convert ATB data from skinny to wider."""
    param_mask = nrelatb.core_metric_parameter.isin(table_unstacker.params)
    # assert not nrelatb[param_mask].duplicated(subset=table_unstacker.idx).any()

    nrelatb_unstacked = (
        nrelatb[param_mask]
        .set_index(table_unstacker.idx)
        .sort_index()[["value"]]
        .drop_duplicates()
        .unstack(level="core_metric_parameter")
    )
    nrelatb_unstacked.columns = nrelatb_unstacked.columns.droplevel()
    return nrelatb_unstacked


@asset
def _core_nrelatb__yearly_projections_by_scenario(
    _core_nrelatb__transform_start,
) -> pd.DataFrame:
    """Transform the yearly NREL ATB scenario table.

    Right now, this just unstacks the table.
    """
    unstacker = Unstacker()
    df = (
        transform_unstack(_core_nrelatb__transform_start, unstacker.scenario_table)
        .pipe(broadcast_fcr_param_across_tech_detail, unstacker.idx_unstacked)
    )
    return df

```

Reply:
i was debating doing this and i agree i do think it makes more sense to do it in the specific table's transform